### PR TITLE
[FLINK-23223] Notifies if there are available data on resumption for pipelined subpartition

### PR DESCRIPTION
**Pull request by gaoyunhaii**
## What is the purpose of the change

This PR tries to fix the issue that when flushAlways is enabled, the data emitted during the channel get blocked might not be able to get notified. 

To fix this issue, we propose to do one additional notify if there is data available when resuming. This piece of data must be emitted after the channel get blocked since the event cause block has been polled, thus the notification would not be repeated (although repeat notification is also not a problem).

There are two cases that might add available data to the channel:
1. There is only one unfinished buffer, and it has been flushed at least once.
2. There is at least one finished buffer. 

When resuming, for the second case, we could detect it according to the buffers in queue, but with the current implementation we could not know if the unfinished buffer is flushed or not. To solve this issue, we could decouple `flushRequested` with `isBlocked`, namely whenever get flushed `flushRequested` will be set to `true` no matter what value `isBlocked` is. This would not cause problem to other parts since we still have checks on `isBlocked` when computing if notify is required and `isAvailable()`. Then when resumed, we could detect if there are available data due to case 1 via `flushRequested`. 

## Brief change log

- 72a7a8aa4ab9a351214f467cb6c8f65d0e3a7384 modifies the logic according to the above proposal. 

## Verifying this change

This change could be verified by the modified UT.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): **no**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: **no**
  - The serializers: **no**
  - The runtime per-record code paths (performance sensitive): **yes** but should not affect the performance
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: **no**
  - The S3 file system connector: **no**

## Documentation

  - Does this pull request introduce a new feature? **no**
  - If yes, how is the feature documented? **not applicable**

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java [hunk 1]
```diff
@@ -350,10 +350,21 @@ BufferAndBacklog pollBuffer() {
...
+
+            // When buffer timeout is 0, there might be data produced during
+            // the channel is blocking and do not notify due to blocking, and
+            // it would not get chance to notify, thus we should notify for it
+            // here.
+            notifyDataAvailable = isDataAvailableUnsafe();
+        }
+
+        if (notifyDataAvailable) {
+            notifyDataAvailable();
```
Comment (wsry) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java:367]: For RemoteInputChannel, the available reader should have been enqueued as available when resuming consumption in PartitionRequestQueue#addCreditOrResumeConsumption method, this data available notification can be duplicated. Maybe we can keep the logic of LocalInputChannel#resumeConsumption method unchanged and only fix the flushRequested flag? What do you think?

**Comment by wsry** [comment]
Hi @gaoyunhaii , thank for investigating and fixing this issue. The code looks good to me. I only have a small comment.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java (hunk 1 above)
Comment (gaoyunhaii) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java:367]: Very thanks pointing out this issue! Sorry that I ignored the logic corresponding to the `RemoteInputChannel`. I think it is reasonable to not introduce the duplicate notification, I'll modify the code.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartitionWithReadViewTest.java [hunk 2]
```diff
@@ -487,42 +487,61 @@ private void testBacklogConsistentWithNumberOfConsumableBuffers(
...
+        checkNumNotificationsAndAvailability(1);
 
+        // Resumption will make the subpartition available.
         resumeConsumptionAndCheckAvailability(Integer.MAX_VALUE, true);
         assertNextBuffer(readView, BUFFER_SIZE, false, 0, false, true);
         assertNextBuffer(readView, BUFFER_SIZE, false, 0, false, true);
     }
 
+    @Test
+    public void testResumeBLockedEmptySubpartition() throws IOException, InterruptedException {
```
Comment (pnowojski) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartitionWithReadViewTest.java:537]: nit `BLocked` -> `Blocked`

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartitionWithReadViewTest.java [hunk 3]
```diff
@@ -487,42 +487,61 @@ private void testBacklogConsistentWithNumberOfConsumableBuffers(
...
         // add a buffer and flush the subpartition
         subpartition.add(createFilledFinishedBufferConsumer(BUFFER_SIZE));
         subpartition.flush();
         // no data available notification after adding a buffer and flushing the subpartition
-        checkNumNotificationsAndAvailability(2);
+        checkNumNotificationsAndAvailability(1);
 
-        resumeConsumptionAndCheckAvailability(Integer.MAX_VALUE, false);
+        // Resumption will make the subpartition available.
+        resumeConsumptionAndCheckAvailability(Integer.MAX_VALUE, true);
```
Comment (pnowojski) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartitionWithReadViewTest.java:515]: Do we have a test case where we resume consumption without previous flush and we assert that the unfinished buffers is NOT available?

**Review by pnowojski — COMMENTED** [review]
LGTM % one maybe missing unit test

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartitionWithReadViewTest.java (hunk 3 above)
Comment (gaoyunhaii) [reply flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartitionWithReadViewTest.java:515]: The test is indeed missed, I'll add one for this case~

**Comment by gaoyunhaii** [comment]
Very thanks @wsry @pnowojski for the reivew! I have updated the PR according to the comments~

**Review by pnowojski — APPROVED** [review]

**Comment by pnowojski** [comment]
Merged manually after squashing. Thanks for the fix @gaoyunhaii

**Pull request closed without merging**